File: LIS.py
# Time Complexity : O(n^2)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        lis = [1] * len(nums)
        max_l = 1
        for i in range(1, len(nums)):
            possible_ext = [1 + lis[j] for j in range(0, i) if nums[j] < nums[i]]
            if len(possible_ext) > 0:
                lis[i] = max(possible_ext)
                max_l = max(max_l, lis[i])
        return max_l

File: lis.java
// Time, space - O(N^2), O(N)

class Solution {
    public int lengthOfLIS(int[] nums) {
        if(nums == null || nums.length == 0) {
            return 0;
        }
        
        int[] dp = new int[nums.length];
        
        Arrays.fill(dp, 1);
        int max = 1;
        for(int i=1;i<nums.length;i++) {
            for(int j=0;j<i;j++) {
                if(nums[i] > nums[j])          {
                    dp[i] = Math.max(dp[i], 1 + dp[j]);
                    max=Math.max(max, dp[i]);
                }       
            }
        }
        return max;
    }
}
